### PR TITLE
184112299 - add gross roi

### PR DESCRIPTION
**Pull request by kamtarnowski**
#### What's this PR do?
PR applies the Gross Yield field. This value is presented on homepage, in Circulating Supply section.

#### How should this be manually tested?
Please verify the gross yield value.

Gross yield is a percentage of total rewards (from last three epochs) divided by the current total active stake.

#### What are the relevant tickets?
[- URL to the PivotalTracker ticket](https://www.pivotaltracker.com/story/show/184112299)

#### Task completed checklist:
- [] Is there appropriate test coverage?
- [] Did I take into consideration possible security vulnerabilities?
- [] Are the controllers secure?
- [] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

File: app/logic/solana_requests_logic.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
```
Comment (kamtarnowski) [inline app/logic/solana_requests_logic.rb:1]: This methods were copied from SolanaLogic module.

**Comment by kamtarnowski** [comment]
The test suit got crashed, since:

It seems we have to update the testing credentials. The mainnet_urls ("https://api.rpcpool.com") responds with 403 error, so we should use the "https://api.mainnet-beta.solana.com" instead.

File: app/services/create_cluster_stats_service.rb [hunk 2]
```diff
@@ -36,6 +38,27 @@ def call
 
   private
 
+  def epochs_annual_roi
+    last_epochs = last_epochs(EPOCHS_TO_CALCULATE)
+    if last_epochs.count > 0
+      epochs_duration = (last_epochs.second.created_at - last_epochs.first.created_at) / 60 / 60 / 24
+      epochs_average_roi = last_epochs(EPOCHS_TO_CALCULATE).map do |epoch|
```
Comment (kbiala) [inline app/services/create_cluster_stats_service.rb:45]: Don't we have `last_epochs(EPOCHS_TO_CALCULATE)` assigned to a variable few lines above?

File: app/services/create_cluster_stats_service.rb [hunk 3]
```diff
@@ -36,6 +38,27 @@ def call
...
+        if epoch.total_rewards && epoch.total_active_stake
+          (epoch.total_rewards / epoch.total_active_stake.to_f) * 100
+        end
+      end.compact.average
+
+      epochs_annual_roi = (365 / epochs_duration) * epochs_average_roi
```
Comment (kbiala) [inline app/services/create_cluster_stats_service.rb:51]: I'd leave both values (epoch duration and year duration) in seconds, or use floats.

File: app/services/create_cluster_stats_service.rb [hunk 4]
```diff
@@ -36,6 +38,27 @@ def call
 
   private
 
+  def epochs_annual_roi
+    last_epochs = last_epochs(EPOCHS_TO_CALCULATE)
+    if last_epochs.count > 0
+      epochs_duration = (last_epochs.second.created_at - last_epochs.first.created_at) / 60 / 60 / 24
```
[the remark this answers is not in the document]
Comment (kbiala) [reply app/services/create_cluster_stats_service.rb:44]: The variable represents the duration of one epoch, so should be called `epochs_duration`, so we do not confuse it with other places where we operate on 3 epochs.

**Review by kbiala — CHANGES_REQUESTED** [review]

File: app/services/create_cluster_stats_service.rb [hunk 5]
```diff
@@ -36,6 +38,25 @@ def call
 
   private
 
+  def epochs_annual_roi
+    if last_epochs.count > 0
+      epoch_duration = (last_epochs.first.created_at - last_epochs.second.created_at)
+      return 0 unless epoch_duration && epoch_duration > 0
```
Comment (kbiala) [inline app/services/create_cluster_stats_service.rb:44]: It should be possible to shorten the condition to `epoch_duration.to_f > 0` or  `epoch_duration.to_i > 0`

File: app/services/total_rewards_update_service.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    rewards_sum
+  end
+
+  private
+
+  def cluster_stat
+    @cluster_stat ||= ClusterStat.find_or_create_by(network: @network)
+  end
+
+  def completed_epoch
```
Comment (kbiala) [inline app/services/total_rewards_update_service.rb:57]: The same comment about relying on ending_slot value.

**Review by kbiala — CHANGES_REQUESTED** [review]

File: app/services/create_cluster_stats_service.rb [hunk 7]
```diff
@@ -36,6 +36,25 @@ def call
...
+    if last_epochs.count > 0
+      epoch_duration = (last_epochs.first.created_at - last_epochs.second.created_at)
+      return 0 unless epoch_duration.to_f > 0
+      epoch_average_roi = last_epochs.map do |epoch|
+        if epoch.total_rewards && epoch.total_active_stake
+          (epoch.total_rewards / epoch.total_active_stake.to_f) * 100
+        end
+      end.compact.average
+
+      epochs_annual_roi = (1.year.to_i / epoch_duration.to_f) * epoch_average_roi
```
Comment (kbiala) [inline app/services/create_cluster_stats_service.rb:49]:
```suggestion
      epochs_annual_roi = (1.year.to_i / epoch_duration.to_f) * epoch_average_roi.to_f
```

**Review by kbiala — CHANGES_REQUESTED** [review]
ClusterStatsWorker should have some retry limit defined.

File: app/services/create_cluster_stats_service.rb (hunk 7 above)
Comment (kbiala) [reply app/services/create_cluster_stats_service.rb:49]: This still need to be fixed. After deploying to production initially this value will be nil and will throw errors.

File: app/javascript/packs/validators/circulating_supply.vue [hunk 8]
```diff
@@ -26,6 +26,9 @@
           <span class="text-muted" v-if="!gross_yield">loading...</span>
           <span v-if="gross_yield">
             <strong class="text-success">{{ gross_yield }}%</strong>
+            <a href="/faq#roi">
```
Comment (kbiala) [inline app/javascript/packs/validators/circulating_supply.vue:29]: Let's move it to the next line

File: app/javascript/packs/validators/circulating_supply.vue [hunk 9]
```diff
@@ -26,6 +26,9 @@
...
+              <i class="fa-solid fa-circle-info font-size-xs text-muted ms-1"></i>
+            </a>
             <div class="small text-muted">
               (Last 3 epochs annualized)
```
Comment (kbiala) [inline app/javascript/packs/validators/circulating_supply.vue:33]:
```suggestion
              (Last 3 epochs annualized)
               <a href="/faq#roi">
                 <i class="fa-solid fa-circle-info font-size-xs text-muted ms-1"></i>
               </a>
```

File: app/models/commission_history.rb [hunk 10]
```diff
@@ -2,17 +2,17 @@
...
-#  created_at             :datetime         not null
-#  updated_at             :datetime         not null
-#  validator_id           :bigint           not null
+#  id                  :bigint           not null, primary key
+#  batch_uuid          :string(191)
+#  commission_after    :float(24)
+#  commission_before   :float(24)
+#  epoch               :integer
+#  epoch_completion    :float(24)
+#  network             :string(191)
```
Comment (kbiala) [inline app/models/commission_history.rb:11]: These changes are not a part of this story.

File: app/views/public/faq.en.html.erb [hunk 11]
```diff
@@ -340,6 +341,20 @@
 
 <hr />
```
Comment (kbiala) [inline app/views/public/faq.en.html.erb:342]: This `<hr />` should go below the whole section, so `Gross ROI` is in the wrong place

File: app/views/public/faq.en.html.erb [hunk 12]
```diff
@@ -340,6 +341,20 @@
...
 
+<h5 id="roi">Gross ROI</h5>
+<p>
+  Gross Return of Interest is a percentage value that represents total return from Solana network to both 
+  validators and stakers. It is calculated as follows:
+</p>
+<pre>
+  (stake_account_rewards + vote_account_rewards / total active stake) * number_of_epochs_in_year
```
Comment (kbiala) [inline app/views/public/faq.en.html.erb:350]:
```suggestion
  (stake_account_rewards + vote_account_rewards / total_active_stake) * number_of_epochs_per_year
```

File: app/views/public/faq.en.html.erb [hunk 13]
```diff
@@ -340,6 +341,20 @@
...
+</pre>
+<p class="mb-5">
+  where stake_account_rewards and vote_account_rewards are the average reward per epoch based on last 3 epochs.
+</p>
```
Comment (kbiala) [inline app/views/public/faq.en.html.erb:354]:
```suggestion
<p>
  Where number_of_epochs_per_year is calculated as follows: seconds_in_year / duration_of_last_epoch_in_seconds. This gives the number of epochs in a year assuming all epochs were as long as the last epoch.
</p>
<p class="mb-5">
  And stake_account_rewards and vote_account_rewards are the average reward per epoch based on last 3 epochs.
</p>
```

**Review by kbiala — CHANGES_REQUESTED** [review]

**Pull request merged**